pandas_gbq/gbq.py

```diff
@@ -577,141 +500,97 @@ def run_query(self, query, **kwargs):
-
-            self.print_elapsed_seconds(
-                '  Got page: {}; {}% done. Elapsed'.format(
-                    len(result_pages),
-                    round(100.0 * current_row / total_rows)))
-
-            if current_row == total_rows:
-                break
-
-            page_token = query_reply.get('pageToken', None)
-
-            if not page_token and current_row < total_rows:
-                raise InvalidPageToken("Required pageToken was missing. "
-                                       "Received {0} of {1} rows"
-                                       .format(current_row, total_rows))
-
-            elif page_token in seen_page_tokens:
-                raise InvalidPageToken("A duplicate pageToken was returned")
-
-            seen_page_tokens.append(page_token)
-
-            try:
-                query_reply = job_collection.getQueryResults(
-                    projectId=job_reference['projectId'],
-                    jobId=job_id,
-                    pageToken=page_token).execute()
-            except HttpError as ex:
-                self.process_http_error(ex)
-
-        if current_row < total_rows:
-            raise InvalidPageToken()
+        try:
+            rows_iter = query_reply.result()
+        except self.http_error as ex:
+            self.process_http_error(ex)
+        result_rows = list(rows_iter)
+        total_rows = rows_iter.total_rows
+        schema = {
+            'fields': [
+                field.to_api_repr()
+                for field in rows_iter.schema],
+        }
 
         # print basic query stats
         self._print('Got {} rows.\n'.format(total_rows))
 
-        return schema, result_pages
+        return schema, result_rows
 
     def load_data(self, dataframe, dataset_id, table_id, chunksize):
-        try:
-            from googleapiclient.errors import HttpError
-        except ImportError:
-            from apiclient.errors import HttpError
+        from google.cloud.bigquery import LoadJobConfig
+        from six import StringIO
 
-        job_id = uuid.uuid4().hex
+        destination_table = self.client.dataset(dataset_id).table(table_id)
+        job_config = LoadJobConfig()
+        job_config.write_disposition = 'WRITE_APPEND'
+        job_config.source_format = 'NEWLINE_DELIMITED_JSON'
         rows = []
         remaining_rows = len(dataframe)
 
         total_rows = remaining_rows
         self._print("\n\n")
 
         for index, row in dataframe.reset_index(drop=True).iterrows():
-            row_dict = dict()
-            row_dict['json'] = json.loads(row.to_json(force_ascii=False,
-                                                      date_unit='s',
-                                                      date_format='iso'))
-            row_dict['insertId'] = job_id + str(index)
-            rows.append(row_dict)
+            row_json = row.to_json(
+                force_ascii=False, date_unit='s', date_format='iso')
+            rows.append(row_json)
```

Comment:
This isn't worse than the last version, but this would be *much* faster if `.to_json` was called on the whole table rather than each row, iterating in python

CSV might be even faster given the reduced space (and pandas can't use nesting or structs anyway). But potentially wait until parquet is GA

Reply:
I'd prefer to keep the current behavior for now and do a subsequent PR for any changes like this for performance improvements. I've filed https://github.com/pydata/pandas-gbq/issues/96 to track the work for speeding up encoding for the `to_gbq()` method.

Reply:
100% re doing that separately